File: bundle/edu.gemini.pot/src/test/scala/edu/gemini/spModel/gemini/gsaoi/PlannedTimeTest.scala
package edu.gemini.spModel.gemini.gsaoi

import org.junit.Test
import org.junit.Assert._
import java.beans.PropertyDescriptor

import edu.gemini.spModel.core.SiderealTarget
import edu.gemini.spModel.data.config.{DefaultParameter, IParameter}
import edu.gemini.spModel.gemini.gems.CanopusWfs
import edu.gemini.spModel.gemini.gsaoi.Gsaoi._
import edu.gemini.spModel.gemini.gsaoi.Gsaoi.Filter._
import edu.gemini.spModel.gemini.seqcomp.SeqRepeatOffset
import edu.gemini.spModel.guide.DefaultGuideOptions
import edu.gemini.spModel.guide.DefaultGuideOptions.Value.{off, on}
import edu.gemini.spModel.obs.plannedtime.{OffsetOverheadCalculator, PlannedTimeCalculator}
import edu.gemini.skycalc.{Angle, Offset}
import edu.gemini.spModel.target.env.GuideProbeTargets
import edu.gemini.spModel.target.SPTarget
import edu.gemini.spModel.target.obsComp.TargetObsComp
import edu.gemini.spModel.target.offset.OffsetPos
import edu.gemini.spModel.test.InstrumentSequenceTestBase
import edu.gemini.spModel.test.InstrumentSequenceTestBase._

import scala.collection.JavaConverters._

class PlannedTimeTest extends InstrumentSequenceTestBase[Gsaoi, GsaoiSeqConfig] {
  override protected def getObsCompSpType = Gsaoi.SP_TYPE
  override protected def getSeqCompSpType = GsaoiSeqConfig.SP_TYPE

  private def getParameter[T](pd: PropertyDescriptor, values: T*): IParameter =
    DefaultParameter.getInstance(pd.getName, values.toList.asJava)

  override def setUp() {
    super.setUp()

    // Add a canopus guide star so that guiding will be turned on
    val env  = getTargetEnvironment
    val grp  = env.getPrimaryGuideGroup
    val target = new SPTarget(SiderealTarget.empty)
    val env2 = env.setPrimaryGuideGroup(grp.put(GuideProbeTargets.create(CanopusWfs.cwfs3, target)))

    val dobj = getTarget.getDataObject.asInstanceOf[TargetObsComp]
    dobj.setTargetEnvironment(env2)
    getTarget.setDataObject(dobj)
  }

  @Test def testFilterWheelMove() {
    val sc0 = createSysConfig
    sc0.putParameter(getParameter(FILTER_PROP, BR_GAMMA, BR_GAMMA))
    setSysConfig(sc0)

    val noMove = PlannedTimeCalculator.instance.calc(getObs).totalTime

    val sc1 = createSysConfig
    sc1.putParameter(getParameter(FILTER_PROP, BR_GAMMA, CH4_LONG))
    setSysConfig(sc1)

    val move = PlannedTimeCalculator.instance.calc(getObs).totalTime

    assertEquals(15000, move - noMove)
  }

  // difference between running the sequence with offset positions (guiding
  // according to the guiding param) vs. with no offset positions
  private def offsetDifference(guiding: Seq[DefaultGuideOptions.Value]): Long = {

    // compute no offset time by creating a sequence with the same number of
    // steps as offset positions
    val sc     = createSysConfig()
    sc.putParameter(getCoaddsParam(Seq.fill(guiding.length)(new java.lang.Integer(1)): _*))
    sc.putParameter(getExpTimeParam(Seq.fill(guiding.length)(new java.lang.Double(1.0)): _*))
    setSysConfig(sc)

    val noOffsetTime = PlannedTimeCalculator.instance.calc(getObs).totalTime

    // Create an offset sequence with a position per guiding option, using
    // the corresponding guiding option.  All the positions are (1,1) away
    // from the previous (with the first one being at (1,1))
    sc.putParameter(getCoaddsParam(1))
    sc.putParameter(getExpTimeParam(1.0))
    setSysConfig(sc)

    val comp     = getInstSeqComp
    val children = comp.getSeqComponents

    comp.setSeqComponents(java.util.Collections.emptyList())

    val offsetComp    = addSeqComponent(comp, SeqRepeatOffset.SP_TYPE)
    val offsetDataObj = offsetComp.getDataObject.asInstanceOf[SeqRepeatOffset]
    val posList       = offsetDataObj.getPosList

    guiding.zipWithIndex foreach { case (g, i) =>
      val pos = posList.addPosition(i+1, i+1)
      pos.setDefaultGuideOption(g)
    }

    offsetComp.setDataObject(offsetDataObj)
    offsetComp.setSeqComponents(children)

    val offsetTime = PlannedTimeCalculator.instance.calc(getObs).totalTime

    offsetTime - noOffsetTime
  }

  // Time for a normal offset of distance (1, 1)
  private val standardOffsetTime = {
    val a0 = new Angle(0, Angle.Unit.ARCSECS)
    val o0 = new Offset(a0, a0)

    val a1 = new Angle(1, Angle.Unit.ARCSECS)
    val o1 = new Offset(a1, a1)

    OffsetOverheadCalculator.instance.calc(o0, o1)
  }

  private def expected(guiding: Seq[DefaultGuideOptions.Value]): Long = {
    val offsets = (off :: guiding.toList).zipWithIndex map { case (guideOption, index) =>
      val pos = new OffsetPos("", index.toDouble, index.toDouble)
      pos.setDefaultGuideOption(guideOption)
      pos
    }

    def isStandardOffset(prev: OffsetPos, cur: OffsetPos): Boolean =
      (prev.getDefaultGuideOption == off) && (cur.getDefaultGuideOption == off)

    val (time, _) = ((0.0,offsets.head)/:offsets.tail) { case ((time,prev),cur) =>
      if (isStandardOffset(prev, cur)) (time + standardOffsetTime, cur)
      else (time + Gsaoi.GUIDED_OFFSET_OVERHEAD, cur)
    }
    (time * 1000).round
  }

  private def validate(guiding: DefaultGuideOptions.Value*) {
    val actual = offsetDifference(guiding)
    assertEquals(expected(guiding), actual)
  }

  @Test def testNoSkyOffsets() {
    validate(on, on)
  }

  @Test def testAllSkyOffsets() {
    validate(off, off)
  }


  @Test def testOneSkyPosition() {
    validate(on, off, on)
  }


  @Test def testMultipleSkyPositions() {
    validate(on, off, off, on)
  }

  @Test def testInitialSkyPositions() {
    validate(off, off, on)
  }
}

File: bundle/edu.gemini.pot/src/test/java/edu/gemini/spModel/gemini/nici/PlannedTimeTest.java
//
// $
//

package edu.gemini.spModel.gemini.nici;

import edu.gemini.pot.sp.ISPSeqComponent;
import edu.gemini.pot.sp.SPComponentType;
import edu.gemini.skycalc.Angle;
import edu.gemini.skycalc.Offset;
import edu.gemini.shared.util.immutable.Pair;
import edu.gemini.shared.util.immutable.Tuple2;
import edu.gemini.spModel.data.config.IParameter;
import edu.gemini.spModel.data.config.ISysConfig;
import static edu.gemini.spModel.gemini.nici.NICIParams.*;

import edu.gemini.spModel.obs.plannedtime.OffsetOverheadCalculator;
import edu.gemini.spModel.obs.plannedtime.PlannedTimeSummaryService;
import edu.gemini.spModel.obs.plannedtime.PlannedTimeSummary;
import edu.gemini.spModel.test.InstrumentSequenceTestBase;
import edu.gemini.spModel.target.offset.OffsetPosList;
import edu.gemini.spModel.seqcomp.SeqRepeatObserve;

import java.beans.PropertyDescriptor;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Test cases for the NICI planned time calculations.
 */
public class PlannedTimeTest extends InstrumentSequenceTestBase<InstNICI, SeqConfigNICI> {


    protected SPComponentType getObsCompSpType() { return InstNICI.SP_TYPE; }

    protected SPComponentType getSeqCompSpType() { return SeqConfigNICI.SP_TYPE; }

    private static IParameter getFocalPlaneMaskParameter(FocalPlaneMask... masks) {
        return getParameter(InstNICI.FOCAL_PLANE_MASK_PROP, masks);
    }

    private static IParameter getChannel1Parameter(Channel1FW... filters) {
        return getParameter(InstNICI.CHANNEL1_FW_PROP, filters);
    }

    private static IParameter getChannel2Parameter(Channel2FW... filters) {
        return getParameter(InstNICI.CHANNEL2_FW_PROP, filters);
    }

    private static <T> IParameter getParameter(PropertyDescriptor pd, T... vals) {
        return getParam(pd.getName(), vals);
    }

    public void testSingleStepExposureTime() throws Exception {
        ISysConfig sc = createSysConfig();
        sc.putParameter(getCoaddsParam(3));
        sc.putParameter(getExpTimeParam(10.0));
        setSysConfig(sc);

        double expected = getInstDataObj().getSetupTime(getObs()).getSeconds() + InstNICI.EXPOSURE_OVERHEAD_CONSTANT;
        expected += 3 * (10.0 + InstNICI.COADD_CONSTANT);
        verify(expected);
    }

    public void testMultiStepExposureTime() throws Exception {
        ISysConfig sc = createSysConfig();
        sc.putParameter(getCoaddsParam(3, 5));
        sc.putParameter(getExpTimeParam(10.0, 12.0));
        setSysConfig(sc);

        // Make two observes at each step.
        SeqRepeatObserve sro = getObserveSeqDataObject();
        sro.setStepCount(2);
        getObserveSeqComp().setDataObject(sro);

        double base = getInstDataObj().getSetupTime(getObs()).getSeconds();

        double exp1 = InstNICI.EXPOSURE_OVERHEAD_CONSTANT + 3 * (10.0 + InstNICI.COADD_CONSTANT);
        double exp2 = InstNICI.EXPOSURE_OVERHEAD_CONSTANT + 5 * (12.0 + InstNICI.COADD_CONSTANT);

        verify(base + 2*(exp1 + exp2));
    }

    public void testDichroicWheelChange() throws Exception {
        IParameter param = getParameter(InstNICI.DICHROIC_WHEEL_PROP, DichroicWheel.H5050_BEAMSPLITTER, DichroicWheel.MIRROR);
        testSingleConfigChange(param);
    }

    public void testFiberCalSourceChange() throws Exception {
        IParameter param = getParameter(InstNICI.FOCS_PROP, Focs.IN_OFF, Focs.IN_ON);
        testSingleConfigChange(param);
    }

    public void testFocalPlaneMaskChange() throws Exception {
        testSingleConfigChange(getFocalPlaneMaskParameter(FocalPlaneMask.MASK_1, FocalPlaneMask.MASK_2));
    }

    public void testPupilImagerChange() throws Exception {
        IParameter param = getParameter(InstNICI.PUPIL_IMAGER_PROP, PupilImager.OPEN, PupilImager.PUPIL_IMAGING);
        testSingleConfigChange(param);
    }

    public void testPupilMaskChange() throws Exception {
        IParameter param = getParameter(InstNICI.PUPIL_MASK_PROP, PupilMask.APODIZED, PupilMask.EIGHTY_PERCENT);
        testSingleConfigChange(param);
    }

    public void testChannel1Change() throws Exception {
        testSingleConfigChange(getChannel1Parameter(Channel1FW.CH4H1L, Channel1FW.CH4H1S));
    }

    public void testChannel2Change() throws Exception {
        testSingleConfigChange(getChannel2Parameter(Channel2FW.CH4H1L, Channel2FW.CH4H1S));
    }

    private void testSingleConfigChange(IParameter param) throws Exception {
        ISysConfig sc = createSysConfig();
        sc.putParameter(param);
        testTwoStepConfigChange(sc);
    }

    // Tests two or more items in the config changing simultaneously.
    public void testMultiConfigChange() throws Exception {
        ISysConfig sc = createSysConfig();
        sc.putParameter(getFocalPlaneMaskParameter(FocalPlaneMask.MASK_1, FocalPlaneMask.MASK_2));
        sc.putParameter(getChannel1Parameter(Channel1FW.CH4H1L, Channel1FW.CH4H1S));
        sc.putParameter(getChannel2Parameter(Channel2FW.CH4H1L, Channel2FW.CH4H1S));
        testTwoStepConfigChange(sc);
    }

    private void testTwoStepConfigChange(ISysConfig sc) throws Exception {
        setSysConfig(sc);

        getInstDataObj().setExposureTime(10.0);
        getInstDataObj().setCoadds(1);
        storeStaticUpdates();

        double base = getInstDataObj().getSetupTime(getObs()).getSeconds();

        // Time for two exposures of 10 seconds with 1 coadd
        double expTime = 2 * (InstNICI.EXPOSURE_OVERHEAD_CONSTANT + 10.0 + InstNICI.COADD_CONSTANT);

        // Setup + exposure + 1 configuration change
        verify(base + expTime + InstNICI.CONFIG_CHANGE_COST);
    }

    private Tuple2<Long, Long> testOffsetCorrection(Offset... offsets) throws Exception {
        ISysConfig sc = createSysConfig();

        // We want the number of steps to be the number of offset positions.
        // We will measure the time required without offset positions for this
        // number of steps.  Later, we will make one instrument iterator step,
        // but with offsets.length offset positions so the step count will be
        // the same and the only difference will be whether offsets were
        // used or not.
        Integer[] coadds = new Integer[offsets.length];
        Arrays.fill(coadds, 1);
        Double[] exptime = new Double[offsets.length];
        Arrays.fill(exptime, 10.0);

        sc.putParameter(getCoaddsParam(coadds));
        sc.putParameter(getExpTimeParam(exptime));
        setSysConfig(sc);

        // Figure out the overhead when no offset positions are given.
        PlannedTimeSummary noOffsetTime = PlannedTimeSummaryService.getTotalTime(getObs());

        // Reduce the number of steps in the inst iterator to 1.
        sc.putParameter(getCoaddsParam(1));
        sc.putParameter(getExpTimeParam(10.0));
        setSysConfig(sc);

        // Now insert offset position(s).
        ISPSeqComponent comp = getInstSeqComp();
        List<ISPSeqComponent> children = comp.getSeqComponents();

        List<ISPSeqComponent> empty = Collections.emptyList();
        comp.setSeqComponents(empty);

        ISPSeqComponent offsetComp = addSeqComponent(comp, SeqRepeatNiciOffset.SP_TYPE);
        SeqRepeatNiciOffset offsetDataObj = (SeqRepeatNiciOffset) offsetComp.getDataObject();
        OffsetPosList<NiciOffsetPos> posList = offsetDataObj.getPosList();
        for (Offset off : offsets) {
            posList.addPosition(off.p().getMagnitude(), off.q().getMagnitude());
        }
        offsetComp.setDataObject(offsetDataObj);
        offsetComp.setSeqComponents(children);

        // Get the time required with the offset position(s) in the mix.
        PlannedTimeSummary offsetTime = PlannedTimeSummaryService.getTotalTime(getObs());

        return new Pair<Long, Long>(noOffsetTime.getExecTime(), offsetTime.getExecTime());
    }

    private void testNoOffsetCorrection(double mag) throws Exception {
        // Get the time required to execute with a big offset.
        Angle a = new Angle(mag, Angle.Unit.ARCSECS);
        Offset offset = new Offset(a, a);
        Tuple2<Long, Long> res = testOffsetCorrection(offset);

        // Should differ by the standard offset overhead.  In other words, no
        // correction should have been applied.
        long actualDiff = res._2() - res._1();

        // Compute the difference in time that we expect to see.
        double secs = OffsetOverheadCalculator.instance.calc(Offset.ZERO_OFFSET, offset);
        long expectedDiff = Math.round(secs * 1000);

        assertEquals(expectedDiff, actualDiff);
    }

    public void testNoOffsetCorrectionWellOverLimit() throws Exception {
        testNoOffsetCorrection(10); // well over small offset limit
    }

    public void testNoOffsetCorrectionJustOverLimit() throws Exception {
        testNoOffsetCorrection(0.22); // just over the small offset limit
    }

    private void testSmallOffsetCorrection(double mag) throws Exception {
        // Get the time required to execute with a small offset.
        Angle a = new Angle(mag, Angle.Unit.ARCSECS);
        Offset offset = new Offset(a, a);
        Tuple2<Long, Long> res = testOffsetCorrection(offset);

        // Should differ by a second, which is the overhead for small offsets.
        long actualDiff = res._2() - res._1();

        assertEquals(1000, actualDiff);
    }

    public void testSmallOffsetCorrectionWellUnderLimit() throws Exception {
        testSmallOffsetCorrection(0.2);      // under the limit
    }

    public void testSmallOffsetCorrectionJustUnderLimit() throws Exception {
        testSmallOffsetCorrection(0.212132); // just under the limit
    }

    public void testUnchangingSmallOffsetCorrection() throws Exception {
        // Get the time required to execute with a small offset.
        Angle a = new Angle(0.2, Angle.Unit.ARCSECS);
        Offset offset = new Offset(a, a);
        Tuple2<Long, Long> res = testOffsetCorrection(offset, offset);

        // Should differ by a second, which is the overhead for small offsets.
        long actualDiff = res._2() - res._1();

        assertEquals(1000, actualDiff);
    }

    public void testChangingSmallOffsetCorrection() throws Exception {
        // Get the time required to execute with a small offset.
        Angle a1 = new Angle(0.2, Angle.Unit.ARCSECS);
        Offset offset1 = new Offset(a1, a1);

        // at the 0.3 limit for a small offset distance
        Angle a2 = new Angle(0.212132, Angle.Unit.ARCSECS);
        Offset offset2 = new Offset(a2, a2);

        Tuple2<Long, Long> res = testOffsetCorrection(offset1, offset2);

        // Should differ by a second for each step, which is the overhead for
        // small offsets.
        long actualDiff = res._2() - res._1();

        assertEquals(2000, actualDiff);
    }

    public void testChangingSmallOffsetCorrectionToNoCorrection() throws Exception {
        // Get the time required to execute with a small offset.
        Angle a1 = new Angle(0.2, Angle.Unit.ARCSECS);
        Offset offset1 = new Offset(a1, a1);

        // Move well away from that offset position.
        Angle a2 = new Angle(-0.22, Angle.Unit.ARCSECS);
        Offset offset2 = new Offset(a2, a2);

        Tuple2<Long, Long> res = testOffsetCorrection(offset1, offset2);

        // Should differ by one second for the first step plus the standard
        // offset overhead.
        long actualDiff = res._2() - res._1();
        long expectedDiff = 1000;
        double secs = OffsetOverheadCalculator.instance.calc(offset2, offset1);
        expectedDiff += Math.round(secs * 1000);

        assertEquals(expectedDiff, actualDiff);
    }
}
